thunder/core/transforms.py

```diff
@@ -1440,16 +1439,24 @@ def grad_func(*args, **kwargs):
         return grad_func
 
     class _GradTransform(Transform):
-        def transform_trace_additionally(self, trc: Trace, *, executors_list: Sequence[Any]) -> Trace:
+        def transform_traces_pre_prologue(
+            self,
+            prologue_trc: Trace,
+            computation_trc: Trace,
+            epilogue_trc: Trace | None,
+            *,
+            executors_list: Sequence[Any],
+        ) -> Trace:
             # Using trc.python_callable() makes it impossible to retrace the
             # function because the python_callable uses python_ctx which replaces
             # symbol occurrences with its symbol._call_ctx function
-            @wraps(trc.python_callable())
+            @wraps(computation_trc.python_callable())
             def python_callable(*args, **kwargs):
-                return eval_trace(trc, *args, **kwargs)
+                computation_trc = dce(computation_trc)
```

Comment:
I would move this up to right at the beginning of `transform_traces_pre_prologue` then you don't need to modify the trace here.